modules: hide implicit modulefiles

xdelaruelle opened the pull request:
When exclude_implicits option is enabled, generate modulefile of implicitly installed software and hide them. Even if implicit, those modulefiles may be referred as dependency in other modulefiles thus they should be generated to make module properly load dependent module.

A new hidden property is added to BaseConfiguration class.

To hide modulefiles, modulercs are generated along modulefiles. Such rc files contain specific module command to indicate a module should be hidden (for instance when using "module avail").

A modulerc property is added to TclFileLayout and LmodFileLayout classes to get fully qualified path name of the modulerc associated to a given modulefile.

Modulerc files will be located in each module directory, next to the version modulefiles. This scheme is supported by both module tool implementations.

modulerc_header and hide_cmd_format attributes are added to TclModulefileWriter and LmodModulefileWriter. They help to know how to generate a modulerc file with hidden commands for each module tool.

Tcl modulerc file requires an header. As we use a command introduced on Modules 4.7 (module-hide --hidden-loaded), a version requirement is added to header string.

Modulerc is created, updated or removed when associated modulefile is written or removed. If an implicit modulefile becomes explicit, hidden command in modulerc for this modulefile is removed. If modulerc becomes empty, this file is removed. Modulerc file is not rewritten when no content change is detected.

Fixes #32013.

xdelaruelle committed:
modules: hide implicit modulefiles

When exclude_implicits option is enabled, generate modulefile of
implicitly installed software and hide them. Even if implicit, those
modulefiles may be referred as dependency in other modulefiles thus they
should be generated to make module properly load dependent module.

A new hidden property is added to BaseConfiguration class.

To hide modulefiles, modulercs are generated along modulefiles. Such rc
files contain specific module command to indicate a module should be
hidden (for instance when using "module avail").

A modulerc property is added to TclFileLayout and LmodFileLayout classes
to get fully qualified path name of the modulerc associated to a given
modulefile.

Modulerc files will be located in each module directory, next to the
version modulefiles. This scheme is supported by both module tool
implementations.

modulerc_header and hide_cmd_format attributes are added to
TclModulefileWriter and LmodModulefileWriter. They help to know how to
generate a modulerc file with hidden commands for each module tool.

Tcl modulerc file requires an header. As we use a command introduced on
Modules 4.7 (module-hide --hidden-loaded), a version requirement is
added to header string.

Modulerc is created, updated or removed when associated modulefile is
written or removed. If an implicit modulefile becomes explicit, hidden
command in modulerc for this modulefile is removed. If modulerc becomes
empty, this file is removed. Modulerc file is not rewritten when no
content change is detected.

Fixes #32013.

xdelaruelle commented:
As purely excluding implicit installation from module generation leads to dependency issue when loading dependent module, `exclude_implicit` option may be renamed `hide_implicit`?

haampie commented:
Yeah, hide implicit is the way 🙃 that's also what I had in mind, but didn't get to.

We could deprecate it but allow both options for now

alalazo commented:
I think we should probably:
1. Rename to `hide_implicit`
2. Have `exclude_implicit` act as an alias and emit a deprecation warning

eugeneswalker commented:
Would it be feasible to change the naming scheme for the hidden, implicit modules? For instance, so that we could generate modules for the explicitly installed packages without hash, but use a hash in the hidden modules for implicitlly installed packages? The motivation here being that it is much easier to disambiguate name clashes in ~140 root specs as in the case of E4S than it is to disambiguate name clashes in all dependencies, which number much greater than 140. We like having no hashes in the explicitly installed packages but it would be hard to go hashless if we generate modules for everything...

haampie commented:
> use a hash in the hidden modules

Sounds reasonable. Maybe the generated config files could also be used to define user friendly aliases for the entrypoints?

xdelaruelle commented:
> Would it be feasible to change the naming scheme for the hidden, implicit modules?

Not sure it would be feasible to have a different naming scheme between implicit and explicit modules. At some point an implicit module can become explicit. Spack would be able to update the modulefile of this installation, but it will not regenerate the dependency information in the other modulefiles.

For this PR, I will focus on the generation of modulerc file to hide implicit modules.

xdelaruelle committed:
modules: rename exclude_implicits into hide_implicits

Rename exclude_implicits option into hide_implicits. Make
exclude_implicits a deprecated option.

simonLeary42 commented:
@haampie Is this feature going to be merged? I would really like to make use of it.

haampie committed:
avoid hiding e.g. mpi in lmod, which can make other explicitly installed modules available after loading

haampie commented:
@simonLeary42 yes, let's get this in.

@xdelaruelle I thought this was merged long ago... apparently not.

Only issue I found was that implicitly installed providers of virtuals would be hidden too, meaning that explicitly installed packages that use say openmpi would never be visible.

So, I relaxed `hidden` for lmod to "not a provider of some virtual and implicitly installed".

Otherwise, this looks good to me!
